### PR TITLE
HDDS-2255. Improve Acl Handler Messages

### PR DESCRIPTION
**Pull request by cxorm**
## What changes were proposed in this pull request?
Add a ```checkAclExist()``` method in ```ObjectStore.java``` and let the method called by ```addAclHandler``` to show the proper messages.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-2255

## How was this patch tested?
Build and check the message.

### CODE REVIEW

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/ObjectStore.java [hunk 1]
```diff
@@ -458,6 +458,23 @@ public boolean addAcl(OzoneObj obj, OzoneAcl acl) throws IOException {
...
+  /**
+   *
+   * @param obj Ozone object for which acl should be checked.
+   * @param acl ozone acl to be checked.
+   * @return true if acl exists in the obj, else false.
+   * @throws IOException if there is error.
+   */
+  public boolean checkAclExist(OzoneObj obj, OzoneAcl acl) throws IOException {
+    for (OzoneAcl existAcl : proxy.getAcl(obj)) {
+      if (existAcl.equals(acl)) {
```
Comment (bharatviswa504) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/ObjectStore.java]: This logic is not completely correct, take a case like USER1:RW, and we are trying to add USER1:R, the acl already exists. equals() will not work in this case.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/web/ozShell/bucket/AddAclBucketHandler.java [hunk 2]
```diff
@@ -88,10 +88,15 @@ public Void call() throws Exception {
             OzoneObj.StoreType.valueOf(storeType))
         .build();
 
-    boolean result = client.getObjectStore().addAcl(obj,
+    boolean aclExisted = client.getObjectStore().checkAclExist(obj,
```
Comment (bharatviswa504) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/web/ozShell/bucket/AddAclBucketHandler.java]: I think we don't need a checkAclExist

addAcl semantic true, acl is added successfully, false means acl already exist that is why we cannot add it. So, no need of checkAclExist()

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/ObjectStore.java (hunk 1 above)
Comment (bharatviswa504) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/ObjectStore.java]: I think from below comment we don't need this new method.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/web/ozShell/keys/AddAclKeyHandler.java [hunk 3]
```diff
@@ -91,10 +91,15 @@ public Void call() throws Exception {
...
 
-    boolean result = client.getObjectStore().addAcl(obj,
+    boolean aclExisted = client.getObjectStore().checkAclExist(obj,
         OzoneAcl.parseAcl(acl));
-
-    System.out.printf("%s%n", "Acl added successfully: " + result);
+    if (aclExisted) {
+      System.out.println("ACL already exists.");
+    } else {
+      boolean result = client.getObjectStore().addAcl(obj,
```
Comment (bharatviswa504) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/web/ozShell/keys/AddAclKeyHandler.java]: I think we don't need a checkAclExist

addAcl semantic true, acl is added successfully, false means acl already exist that is why we cannot add it. So, no need of checkAclExist()

Comment (cxorm) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/web/ozShell/keys/AddAclKeyHandler.java]: Yes, thanks @bharatviswa504 for the comment.

And I found that if the acl already exist when we add it, the ```OMException``` would be thrown by the ```OmOzoneAclMap#addAcl``` result in returning false.
So we don't need ```checkAclExist```.

Fixed.

**Comment by cxorm** [comment]
Thanks @bharatviswa504 for the comment.

With the trace of SetAclHanlder, 
we found that the Set operation would clear AclList, 
so the result will be true.

And RemoveAclHandler Updated.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/web/ozShell/keys/AddAclKeyHandler.java (hunk 3 above)
Comment (cxorm) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/web/ozShell/keys/AddAclKeyHandler.java]: Thanks @bharatviswa504 

Sorry for last comment.
I found that

```ozone sh vol setacl --acls=user:newUser:wr,user:newUser:w,user:newUser:r /vol``` 
will return ```false``` for the already existed acl.

```ozone sh bucket setacl --acls=user:newUser:wr,user:newUser:w,user:newUser:r /vol/buc```
will return ```true```

```ozone sh key setacl --acls=user:newUser:wr,user:newUser:w,user:newUser:r /vol/buc/ikey```
will return ```true```

So I update the ```SetAclVolumeHandler```.

Could we fix the ```SetAclVolumeHandler``` &```SetAclVolumeHandler``` to show the right message for the existed acl ?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/web/ozShell/bucket/AddAclBucketHandler.java [hunk 4]
```diff
@@ -91,7 +91,11 @@ public Void call() throws Exception {
     boolean result = client.getObjectStore().addAcl(obj,
         OzoneAcl.parseAcl(acl));
 
-    System.out.printf("%s%n", "Acl added successfully: " + result);
+    String message = result
+        ? ("Acl added successfully: ") + result
```
Comment (bharatviswa504) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/web/ozShell/bucket/AddAclBucketHandler.java]: Can we change this message if true: Acl added successfully, in false, Acl already exists.
I think in false case printing Acl added successfully: false Acl already exists which is not great in my suggestion.

**Comment by bharatviswa504** [comment]
> Thanks @bharatviswa504 for the comment.
> 
> `RemoveAclHandler` updated.
> 
> And I found that
> 
> `ozone sh vol setacl --acls=user:newUser:wr,user:newUser:w,user:newUser:r /vol` will return `false` for the already existed acl.
> 
> `ozone sh bucket setacl --acls=user:newUser:wr,user:newUser:w,user:newUser:r /vol/buc` will return `true`
> 
> `ozone sh key setacl --acls=user:newUser:wr,user:newUser:w,user:newUser:r /vol/buc/ikey` will return `true`
> 
> So we update the `SetAclVolumeHandler`.
> 
> Could we fix the `SetAclBucketHandler#setAcl()` & `SetAclKeyHandler#setAcl()` to show the right message for the existed acl ?

OzoneAclUtil.java
```
 public static boolean setAcl(List<OzoneAcl> existingAcls,
      List<OzoneAcl> newAcls) {
    if (existingAcls == null) {
      return false;
    } else {
      existingAcls.clear();
      if (newAcls != null) {
        existingAcls.addAll(newAcls);
      }
    }
    return true;
  }
```

setAcl logic used by Bucket/Volume/Key SetAcl. if new Acls are being set we just override and return true.  and top existingAcls == null will mostly not happen, as we set in builder like below.

OmBucketInfo.java L256:
```
    public Builder() {
      //Default values
      this.acls = new ArrayList<>();
      this.isVersionEnabled = false;
      this.storageType = StorageType.DISK;
      this.metadata = new HashMap<>();
    }
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/web/ozShell/bucket/AddAclBucketHandler.java (hunk 4 above)
Comment (cxorm) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/web/ozShell/bucket/AddAclBucketHandler.java]: Yes, I also think this message(```true``` : Acl added successfully.  ```false```: Acl already exists.)
would be more readable for developer and user.

**Comment by cxorm** [comment]
> > Thanks @bharatviswa504 for the comment.
> > `RemoveAclHandler` updated.
> > And I found that
> > `ozone sh vol setacl --acls=user:newUser:wr,user:newUser:w,user:newUser:r /vol` will return `false` for the already existed acl.
> > `ozone sh bucket setacl --acls=user:newUser:wr,user:newUser:w,user:newUser:r /vol/buc` will return `true`
> > `ozone sh key setacl --acls=user:newUser:wr,user:newUser:w,user:newUser:r /vol/buc/ikey` will return `true`
> > So we update the `SetAclVolumeHandler`.
> > Could we fix the `SetAclBucketHandler#setAcl()` & `SetAclKeyHandler#setAcl()` to show the right message for the existed acl ?
> 
> OzoneAclUtil.java
> 
> ```
>  public static boolean setAcl(List<OzoneAcl> existingAcls,
>       List<OzoneAcl> newAcls) {
>     if (existingAcls == null) {
>       return false;
>     } else {
>       existingAcls.clear();
>       if (newAcls != null) {
>         existingAcls.addAll(newAcls);
>       }
>     }
>     return true;
>   }
> ```
> 
> setAcl logic used by Bucket/Volume/Key SetAcl. if new Acls are being set we just override and return true. and top existingAcls == null will mostly not happen, as we set in builder like below.
> 
> OmBucketInfo.java L256:
> 
> ```
>     public Builder() {
>       //Default values
>       this.acls = new ArrayList<>();
>       this.isVersionEnabled = false;
>       this.storageType = StorageType.DISK;
>       this.metadata = new HashMap<>();
>     }
> ```
> 
> So, I have not got how we return false for already existing acl.

Thanks @bharatviswa504 for the review.

Yes, the Bucket/Key ```SetAcl``` use the ```OzoneAclUtil#setAcl```, and it will return ```ture``` mostly.

But the Volume ```Setacl``` use the ```OmOzoneAclMap#setAcl``` that call the ```add``` which causes ```OMException``` when adding the existing acl, and the ```OMException``` would result in ```false```.

A command would return ```false``` may like: 
```ozone sh vol setacl --acls=user:newUser:wr,user:newUser:w,user:newUser:r /vol```

**Comment by bharatviswa504** [comment]
Below are the find bugs that need to be fixed Jenkins reported.
M D DLS: Dead store to result in org.apache.hadoop.ozone.web.ozShell.bucket.SetAclBucketHandler.call()  At SetAclBucketHandler.java:[line 91]
M D DLS: Dead store to result in org.apache.hadoop.ozone.web.ozShell.keys.SetAclKeyHandler.call()  At SetAclKeyHandler.java:[line 93]

**Comment by cxorm** [comment]
> Below are the find bugs that need to be fixed Jenkins reported.
> M D DLS: Dead store to result in org.apache.hadoop.ozone.web.ozShell.bucket.SetAclBucketHandler.call() At SetAclBucketHandler.java:[line 91]
> M D DLS: Dead store to result in org.apache.hadoop.ozone.web.ozShell.keys.SetAclKeyHandler.call() At SetAclKeyHandler.java:[line 93]
> 
> Unused variable result in SetAclBucketHandler and SetAclKeyHandler.

Thanks @bharatviswa504 for the comment.
Fixed

**Comment by bharatviswa504** [comment]
Thank You @cxorm for the update.
I have committed this to the trunk.

**Comment by cxorm** [comment]
Thanks @bharatviswa504  for the review and merge.

**Pull request merged**